docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+# Overview
+Currently cascading deletion is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present two alternative designs. [Design #1](#design-1-tombstone--gc) uses the tombstone mechanism, and [Desig #2](#design-2-finalizer-framework--gc) uses the finalizer mechanism.
+
+# Goals
+* Supporting cascading deletion at the server-side. 
+* Centralizing the cascading deletion logic, rather than spreading in controllers.
+* Allowing optionally orphan the dependent objects.
+
+# Non goals
+* Releasing the name of an object immediately, so it can be reused ASAP.
+
+# Design 1. Tombstone + GC
+## API Changes
+```
+type DeleteOptions struct {
+	…
+	Orphaning bool
+}
+```
+**DeleteOptions.Orphaning**: allows a user to express whether the child objects should be orphaned.
+
+```
+type ObjectMeta struct {
+	...	
+	ParentReferences []ObjectReference
```

Comment:
I would like us to add `ObjectMeta.NamespaceUid` as well so the namespace controller can do a similar reaping when orphans sneak out the window during namespace deletion.